app/validators/rules/rule_validator.py

```diff
@@ -0,0 +1,207 @@
+        f"Argument one of the `map` operator does not reference `{SELF}`"
+    )
+    SELF_REFERENCE_OUTSIDE_MAP_OPERATOR = (
+        f"Reference to {SELF} was made outside of the `map` operator"
+    )
+
+    def __init__(self, rules, origin_id, questionnaire_schema):
+        super().__init__(rules)
+        self.rules = rules
+        self.questionnaire_schema = questionnaire_schema
+        self.context["origin_id"] = origin_id
+
+    def validate(self):
+        self._validate_rule(self.rules)
+        return self.errors
+
+    def _validate_rule(self, rules, *, allow_self_reference=False):
+        operator_name = next(iter(rules))
+        allow_self_reference = allow_self_reference or operator_name == Operator.MAP
+
+        arguments = rules[operator_name]
+        for argument in arguments:
+            if isinstance(argument, dict) and any(
+                operator in argument for operator in ALL_OPERATORS
+            ):
+                self._validate_rule(argument, allow_self_reference=allow_self_reference)
+
+        if operator_name == Operator.DATE:
+            self._validate_date_operator(rules)
+
+        elif operator_name == Operator.COUNT:
+            self._validate_count_operator(rules)
+
+        elif operator_name == Operator.MAP:
+            self._validate_map_operator(rules)
+
+        if operator_name in [Operator.FORMAT_DATE, Operator.DATE]:
+            self._validate_self_references(
+                rules, operator_name, allow_self_reference=allow_self_reference
+            )
+
+        self._validate_options(rules, operator_name)
+
+    def _validate_self_references(self, rules, operator_name, *, allow_self_reference):
+        """
+        Validate references to `self` are within the context of the `map` operator
+        """
+        arguments_for_non_map_operators = self._get_flattened_arguments_for_non_map_operators(
+            rules[operator_name]
+        )
+        if SELF in arguments_for_non_map_operators and not allow_self_reference:
+            self.add_error(self.SELF_REFERENCE_OUTSIDE_MAP_OPERATOR, rule=rules)
+
+    def _validate_map_operator(self, operator):
+        """
+        Validates that the first argument to the map operator has a reference to `self`
+        The second argument is currently not validated here as it can currently only be `date-range`
+        """
+        function_to_map_over_arguments = list(operator["map"][0].values())[0]
+        if SELF in function_to_map_over_arguments:
+            return None
+
+        arguments_for_non_map_operators = self._get_flattened_arguments_for_non_map_operators(
+            function_to_map_over_arguments
+        )
+
+        if SELF not in arguments_for_non_map_operators:
+            self.add_error(
+                self.MAP_OPERATOR_WITHOUT_SELF_REFERENCE, rule=operator["map"][0]
+            )
+
+    def _get_flattened_arguments_for_non_map_operators(self, arguments):
+        non_operator_arguments = []
```

Comment:
We seem inconsistent with doc strings in this class

Reply:
Have added a doc string for this.